srtcore/core.cpp

```diff
@@ -10909,6 +10889,68 @@ void CUDT::checkTimers()
     }
 }
 
+void CUDT::updateBrokenConnection(int errorcode)
+{
+    releaseSynch();
+
+    int token = -1;
+    bool pending_broken = false;
+
+#if ENABLE_EXPERIMENTAL_BONDING
+    {
+        ScopedLock guard_group_existence (s_UDTUnited.m_GlobControlLock);
+        if (m_parent->m_IncludedGroup)
+        {
+            token = m_parent->m_IncludedIter->token;
+            if (m_parent->m_IncludedIter->sndstate == SRT_GST_PENDING)
+            {
+                HLOGC(gmlog.Debug, log << "updateBrokenConnection: a pending link was broken - will be removed");
+                pending_broken = true;
+            }
+            else
+            {
+                HLOGC(gmlog.Debug, log << "updateBrokenConnection: state=" << CUDTGroup::StateStr(m_parent->m_IncludedIter->sndstate) << " a used link was broken - not closing automatically");
+            }
+
+            m_parent->m_IncludedIter->sndstate = SRT_GST_BROKEN;
+            m_parent->m_IncludedIter->rcvstate = SRT_GST_BROKEN;
+        }
+    }
+#endif
+
+    // app can call any UDT API to learn the connection_broken error
+    s_UDTUnited.m_EPoll.update_events(m_SocketID, m_sPollID, SRT_EPOLL_IN | SRT_EPOLL_OUT | SRT_EPOLL_ERR, true);
+    CGlobEvent::triggerEvent();
+    if (m_cbConnectHook)
+    {
+        CALLBACK_CALL(m_cbConnectHook, m_SocketID, errorcode, m_PeerAddr.get(), token);
+    }
```

Comment:
Maybe?

```suggestion
    if (m_cbConnectHook && m_bConnecting)
    {
        CALLBACK_CALL(m_cbConnectHook, m_SocketID, errorcode, m_PeerAddr.get(), token);
    }
```

Reply:
Actually the most important intention of the callback is to be called at the moment when the connection is broken. So this would be wrong.

Reply:
The docs on [srt_connect_callback](https://github.com/Haivision/srt/blob/master/docs/API-functions.md#srt_connect_callback) disagree:
> This call installs a callback hook, which will be executed on a given clr socket just after the pending connection situation in the background has been resolved (that is, when the connection succeeded or failed).

See also [my comment above](https://github.com/Haivision/srt/pull/1639#discussion_r518659096).

Reply:
Ok, I was kinda unclear.

The intention of the callback is to be called at the moment when the connection is SUCCESSFUL OR BROKEN. That's exactly what is written in the documentation you cited.

Whereas your proposal limits the cases of calling it to successful resolution only.

Reply:
The problem I am raising is the following.
Connection hook will now be **also** called at shutdown `UMSG_SHUTDOWN`. 

`SHUTDOWN` may arrive:

1. right after HS response, when **callback makes sense**,
2. at the end of transmission if the peer closes the connection, then **callback does not look reasonable**.

What do you think about situation no. 2?

On my proposal, after connection is established, `m_bConnecting` becomes false. The question is if the callback is called befare of after this?

Reply:
The situation 2 was the exact reason for introducing the connect callbacks. It was in order to allow a non-blocking-style function to notify this information so that the application knows that it must revive the link.